src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,311 @@
+from flint.flintlib.fmpz cimport (
+    fmpz_t,
+    fmpz_one,
+    fmpz_set,
+    fmpz_init,
+    fmpz_clear,
+    fmpz_equal
+)
+from flint.flintlib.fmpz_mod cimport *
+
+from flint.utils.typecheck cimport typecheck
+from flint.flint_base.flint_base cimport flint_scalar
+from flint.types.fmpz cimport (
+    fmpz,
+    any_as_fmpz,
+    fmpz_get_intlong
+)
+
+
+cdef class fmpz_mod_ctx:
+    """
+    """
+    def __cinit__(self, mod):
+        """
+        """
+        # Ensure modulus is fmpz type
+        if not typecheck(mod, fmpz):
+            mod = any_as_fmpz(mod)
+            if mod is NotImplemented:
+                raise TypeError("Context modulus must be able to be case to an `fmpz` type")
+
+        # Ensure modulus is positive
+        if mod < 1:
+            raise ValueError("Modulus is expected to be positive")
+
+        # Init the context
+        fmpz_mod_ctx_init(self.val, (<fmpz>mod).val)
+
+    def __dealloc__(self):
+        fmpz_mod_ctx_clear(self.val)
```

Comment:
In the same way that everything in Flint has an `_init` method and a `_clear` method every Cython type that has some C-level data needs to have `__cinit__` and `__dealloc__`. The only thing that should really be done here is just that `__cinit__` should call the Flint `_init` function to ensure that the C level data attached to the Cython class instance is in a consistent state ready for when `__dealloc__` is called and will call the `_clear` function.

Note that `__dealloc__` will still be called if this function exits with an exception so this function must not exit unless `fmpz_mod_ctx_clear` can be called safely right after. It should be considered an error to call the `_clear` function if the `_init` function was not called so any exit from `__cinit__` needs to have called `fmpz_mod_ctx_init`.

Probably what should be done here is to use `fmpz_mod_ctx_init` but set the modulus to 1. Then I guess the `__init__` method can call `fmpz_mod_ctx_set_modulus` to change it to the appropriate value. Or perhaps what is needed is just to set the modulus to one before doing any type checks and then change it if an exception is not going to be raised.

Reply:
https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#initialisation-methods-cinit-and-init

Reply:
For now I will init the modulus to one in `cinit` and then move everything back to `__init__`.